objects/src/accounts/builder/mod.rs

```diff
@@ -0,0 +1,436 @@
+                        Some(Box::new(err)),
+                    )
+                },
+            )?;
+
+        Ok((init_seed, vault, code, storage))
+    }
+
+    /// Grinds a new [`AccountId`] using the `init_seed` as a starting point.
+    fn grind_account_id(
+        &self,
+        init_seed: [u8; 32],
+        code_commitment: Digest,
+        storage_commitment: Digest,
+    ) -> Result<(AccountId, Word), AccountError> {
+        let seed = AccountId::get_account_seed(
+            init_seed,
+            self.account_type,
+            self.storage_mode,
+            code_commitment,
+            storage_commitment,
+        )
+        .map_err(|err| {
+            AccountError::BuildError("account seed generation failed".into(), Some(Box::new(err)))
+        })?;
+
+        let account_id = AccountId::new(seed, code_commitment, storage_commitment)
+            .expect("get_account_seed should provide a suitable seed");
+
+        Ok((account_id, seed))
+    }
+
+    /// Builds an [`Account`] out of the configured builder.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if:
+    /// - The init seed is not set.
+    /// - Any of the components does not support the set account type.
+    /// - The number of procedures in all merged components is 0 or exceeds
+    ///   [`AccountCode::MAX_NUM_PROCEDURES`](crate::accounts::AccountCode::MAX_NUM_PROCEDURES).
+    /// - Two or more libraries export a procedure with the same MAST root.
+    /// - The number of [`StorageSlot`](crate::accounts::StorageSlot)s of all components exceeds
+    ///   255.
+    /// - [`MastForest::merge`](vm_processor::MastForest::merge) fails on the given components.
+    /// - If duplicate assets were added to the builder (only under the `testing` feature).
+    /// - If the vault is not empty on new accounts (only under the `testing` feature).
+    pub fn build(self) -> Result<(Account, Word), AccountError> {
+        let (init_seed, vault, code, storage) = self.build_inner()?;
+
+        let (account_id, seed) =
+            self.grind_account_id(init_seed, code.commitment(), storage.commitment())?;
+
+        debug_assert_eq!(account_id.account_type(), self.account_type);
+        debug_assert_eq!(account_id.storage_mode(), self.storage_mode);
+
+        let account = Account::from_parts(account_id, vault, storage, code, self.nonce);
+
+        Ok((account, seed))
+    }
+
+    /// The build method optimized for testing scenarios. The only difference between this method
+    /// and the [`Self::build`] method is that when building existing accounts, this function
+    /// returns `None` for the seed, skips the grinding of an account id and constructs one
+    /// instead. Hence it is always preferable to use this method in testing code.
+    ///
+    /// For possible errors, see the documentation of [`Self::build`].
+    #[cfg(feature = "testing")]
+    pub fn build_testing(self) -> Result<(Account, Option<Word>), AccountError> {
```

Comment:
Should this be moved into the impl block on line 210 below (to put all testing code together)?

Also, one other potential way to handle the separation between testing/non-testing code is to rename this into `build_existing` so that it is used for constructing only existing accounts. Then, we can omit `Option<Word>` from the return type. The simple `build()` would then be used for constructing only new accounts.

We can even modify the signature of `build_existing()` to be something like:
```rust
pub fn build_existing(self, nonce: Felt) -> Result<Account, AccountError>
```
And then we can get rid of the `nonce` field in the struct. But I'm also fine keeping the pure builder pattern (i.e., with the `nonce` field).

Reply:
Sure, yeah, moved it to the `cfg`'d `impl` block.

I also considered using `build_existing` first but settled on `build_testing` to accommodate the way we currently use it in test code, two examples:

- Here we pass the account_seed directly to the `TransactionContextBuilder` so using an `Option` is convenient.
https://github.com/0xPolygonMiden/miden-base/blob/bc8a30ba44f26342d6dc48bef99eef012bd9d1d2/miden-tx/src/tests/kernel_tests/test_prologue.rs#L534-L557
- In the MockChain we may call this method with a builder configured to build a  new or existing account, so having a single method handle both is useful since that method is unaware of whether it is buildng a new or existing account. If we wanted to use `build_existing` we'd have to restructure it and it seemed easier to use `build_testing` instead.
https://github.com/0xPolygonMiden/miden-base/blob/bc8a30ba44f26342d6dc48bef99eef012bd9d1d2/miden-tx/src/testing/mock_chain.rs#L374-L401

I think I would keep the `nonce` method as-is then too, since it only makes sense with `build_existing`.

Reply:
> Here we pass the account_seed directly to the `TransactionContextBuilder` so using an `Option` is convenient.

Here we are building a new account - right? So, shouldn't we be using `build()` method?

> In the MockChain we may call this method with a builder configured to build a  new or existing account, so having a single method handle both is useful since that method is unaware of whether it is buildng a new or existing account. If we wanted to use `build_existing` we'd have to restructure it and it seemed easier to use `build_testing` instead.

Yeah - in this case we don't know which one to call - but also, this method is used only a handful of times - so, I wonder if we should just refactor it.